### ISSUE TITLE
TuesPechkin unable to load DLL 'wkhtmltox.dll'

### ISSUE THREAD

**Issue by deliriousDev** [tuespetre/TuesPechkin#57]
I've been using TuesPechkin for some time now and today I went to update the nuget package to the new version 2.0.0+ and noticed that Factory.Create() no longer resolved, so I went to read on the GitHub the changes made and noticed it now expects the path to the dll?

IConverter converter =
new ThreadSafeConverter(
new PdfToolset(
new StaticDeployment(DLL_FOLDER_PATH)));
For the past few hours I've tried almost all the paths I can think of, "\bin", "\app_data", "\app_start", etc and I can't seem to find or figure out what it wants for the path and what dll?

I can see the TuesPechkin dll in my bin folder and it was the first path I tried, but I got the following error:

Additional information: Unable to load DLL 'wkhtmltox.dll': The specified module could not be found. (Exception from HRESULT: 0x8007007E)

Where is that dll and now can I get it as the library doesn't seem to contain it, I tried installing the TuesPechkin.Wkhtmltox.Win32 package but the dll still is nowhere to be found. Also I am using this in a asp.net website project so I assume that using the following should work for obtaining the path, right?

var path = HttpContext.Current.Server.MapPath(@"~\bin\TuesPechkin.dll");

http://stackoverflow.com/questions/27882601/tuespechkin-unable-to-load-dll-wkhtmltox-dll

**Comment by tuespetre** [comment]
The IDeployment represents the folder where the wkhtmltox.dll file(s) reside, or in the case of using Win32/Win64 embedded libraries, the folder where it will drop the file(s). Just use Path.GetTempPath() or something like that.

**Comment by deliriousDev** [comment]
When I try using Path.GetTempPath() is can't find the file. I ended up copying the wkhtmltox.dll DLL into the root of my project and setting it to "copy always", but I am just not sure if that is the best way to do it. 

the wkhtmltox.dll is nowhere in the TuesPechkin nuget package from what I could see, I did a nuget search and installed the Win32 library for wkhtmltox.dll and that didn't work either. 

I ended up looking through the TuesPechkin repo and inside the Tests project I saw the dll and downloaded it and added it as I said above to the project.

Why did this change? In the older version you didn't need to bother with any DLL folder path, it just worked. :) Anyways I did manage to figure it out but the documentation around the whole DLL folder could be improved.

**Comment by tuespetre** [comment]
In the past everything was concealed behind that Factory class; everything has now been broken down into interfaces to open it up for extension. I think I will include a TempPathDeployment class in next week's release to imitate the old behavior. :)

The Win32 one might also not be the one for you; are you sure you wouldn't be using the Win64 package instead?

- Derek Gray

**Comment by tuespetre** [comment]
Also, please see the test web app project in source for a usage example.

- Derek Gray

**Comment by deliriousDev** [comment]
It would be great if the library had a way to not need to provide a DLL path. The Test web app project was quite helpful, perhaps adding a bit more context or usage examples from it to the documentation could help others. I managed to figure it out but I'll admit it was frustrating spending the amount of time I did on such a simple thing, thank god you had a test app to look at... put that down as yet another benefit of TDD lol. Thanks tuespete for the quick response and I look forward to the next update. 

Do you know if 2.0 has better CSS3 and Font support? I noticed in the old one it didn't render text with borders very well.

**Comment by tuespetre** [comment]
The CSS3 and font support is out of my hands but as I said I will create a `TempFolderDeployment` class for ease of use and add it to the documentation.

**Comment by deliriousDev** [comment]
Just to clarify if I understand you correctly. 2.1.0 you will just need to 

1) Install-Package TuesPechkin
2) Install-Package TuesPechkin.Wkhtmltox.Win32
3) Use TempFolderDeployment in your converter

I tried to add the DLL to my asp.net project root and it worked fine until I uploaded to my host and then it said it couldn't find the DLL, even though it was in the project of the website. 

Sorry, I am new to a lot of this. I love TuesPechkin and the work you've done with it. I appreciate your help alot. Thank you.

**Comment by tuespetre** [comment]
If 32-bit is the right version for you, then yes, the following code for 2.1.0 will essentially perform the same as 1.0.3:

```
IConverter converter =
    new ThreadSafeConverter(
        new RemotingToolset<PdfToolset>(
            new Win32EmbeddedDeployment(
                new TempFolderDeployment())));

// Keep the converter somewhere static, or as a singleton instance!

byte[] result = converter.convert(document);
```

**Comment by deliriousDev** [comment]
Any ETA for when 2.1.0 will be on Nuget? Thank you for getting this out so fast and helping us out with the issue, much appreciated.

**Comment by deliriousDev** [comment]
I updated the DLL and it looks to be running but it's hanging?

        public byte[] GetPdfBytes(string url)
        {
            var pdfAsByte = new byte[0];
            if (!string.IsNullOrEmpty(url))
            {
                var htmlToPdfDoc = new HtmlToPdfDocument()
                {
                    GlobalSettings = new GlobalSettings()
                    {
                        DocumentTitle = "Order Confirmation",
                        PaperSize = System.Drawing.Printing.PaperKind.Letter,
                        Margins =
                        {
                            All = 1.375,
                            Unit = Unit.Centimeters
                        }
                    },
                    Objects = 
                    {
                        new ObjectSettings
                        {
                            PageUrl = url
                        }
                    }
                };

                IConverter converter = new ThreadSafeConverter(new RemotingToolset<PdfToolset>(new Win32EmbeddedDeployment(new TempFolderDeployment()))); 
                pdfAsByte = converter.Convert(htmlToPdfDoc);
            }
            return pdfAsByte;
        }

**Comment by tuespetre** [comment]
I'm looking at releasing 2.1.0 (as well as the 0.12.2 embedded versions) this weekend.

The only thing I see wrong with your example is that you are instantiating a converter every time; if you are using a ThreadSafeConverter it must be stored somewhere static, or the same instance needs to be passed in via injection (whichever way floats your boat; it will still be somewhere static.)

**Comment by deliriousDev** [comment]
I'm using Ninject so I'll register the IConverter and DI it. Thanks.

**Comment by deliriousDev** [comment]
Ok I set it to a static variable;

        private static IConverter converter =
            new ThreadSafeConverter(
                new RemotingToolset<PdfToolset>(
                    new Win32EmbeddedDeployment(
                        new TempFolderDeployment())));

Runs perfectly on my localhost, but when deployed to IIS on GoDaddy it hangs....

**Comment by tuespetre** [comment]
These changes are now on NuGet as 2.1.0. I have also been made aware of a dependency on the Visual C++ runtime for the Embedded deployments (as they are now.) Do you have Visual C++ runtime installed on the machine you're having issues with?

**Comment by deliriousDev** [comment]
The hanging issues appear on both GoDaddy and HostGator hosting accounts, from what I could find it appears they should have the C++ runtime installed.

I ended up rolling back to the previous version which still hangs but not as often.

**Comment by muxdaddy** [comment]
First off, nice work.  Using it locally is awesome and I can't wait to show off what I've done with it locally to some others.

But..., I'm also having troubles upon deploying it to our staging server.  It has the Visual C++ runtime installed.  I've tried it numerous ways and they all come back with this same error.

I've tried the Win64EmbeddedDeployment with Static (to the bin AND the GetTempPath()) and the TempFolderDeployment.

Here's where I'm at now after downloading the latest wkhtmltox.dll and just putting it in the folder listed below.

``` c#
        private static string wkhtmltoxPath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "/bin/win64");
        private static IConverter converter = new ThreadSafeConverter(new RemotingToolset<PdfToolset>(new StaticDeployment(wkhtmltoxPath)));
```

Any thoughts?

**Comment by tuespetre** [comment]
Which error is it again? Not finding the dll?

**Comment by muxdaddy** [comment]
Yeah,

Unable to load dll 'wkhtmltox.dll': The specified module could not be found. 
(Exception from HRESULT: 0x8007007E)

I'm using 0.12.2.1 of wkhtmltox.dll and 2.1.0 of yours..

**Comment by tuespetre** [comment]
I see. Please update to 2.1.1, it addresses this issue.

**Comment by muxdaddy** [comment]
Hmmm, didn’t seem to have an effect…, same error.  I rebooted the machine, just to be safe…

**Comment by tuespetre** [comment]
Maybe if I had scrolled over in your example I would have seen that you are currently using the StaticDeployment by itself... 2.1.1 just addressed an issue with the EmbeddedDeployment. I've not been able to reproduce any issues with the StaticDeployment by itself; I suspect something is going on with how your staging environment is configured.

I would give the EmbeddedDeployment another shot now with 2.1.1.

**Comment by muxdaddy** [comment]
EmbeddedDeployment using Temp or Static for where it should place the wkhtmlox.dll… OR doesn’t it matter, they should both work?

Thanks for your help, btw :)

**Comment by tuespetre** [comment]
I would recommend the TempFolderDeployment.

**Comment by muxdaddy** [comment]
What would I look for in my staging server config?  They are ours, so I don’t have to guess at anything…  The following is bombing just the same with 2.1.1 and 12.2.1 of the Win64, both fresh from NuGet

private static IConverter converter = new ThreadSafeConverter(new RemotingToolset<PdfToolset>(new Win64EmbeddedDeployment(new TempFolderDeployment())));

**Comment by tuespetre** [comment]
Hmmm... I created a fresh, empty web application with a Default.aspx page with this code behind:

```
namespace TuesPechkinTest
{
    public partial class WebForm1 : System.Web.UI.Page
    {
        static IConverter converter =
            new ThreadSafeConverter(
                new RemotingToolset<PdfToolset>(
                    new Win64EmbeddedDeployment(
                        new TempFolderDeployment())));

        protected void Page_Load(object sender, EventArgs e)
        {
            var result = converter.Convert(new HtmlToPdfDocument { Objects = { new ObjectSettings { PageUrl = "www.google.com" } } });

            Response.Clear();
            Response.ClearHeaders();
            Response.ClearContent();
            Response.AddHeader("Content-Disposition", "attachment; filename=test.pdf");
            Response.AddHeader("Content-Length", result.Length.ToString());
            Response.ContentType = "text/plain";
            Response.Flush();

            new MemoryStream(result).CopyTo(Response.OutputStream);

            Response.End();
        }
    }
}
```

Running the application produces a copy of wkhtmltox.dll in this folder: `C:\Users\grayd\AppData\Local\Temp\-489600926\8\0.12.2.1` and I receive a pdf with Google's homepage printed in it.

Are you able to see where in the temp folder the dll is being deployed? If/when you can find it, what is the full path of the folder?

**Comment by muxdaddy** [comment]
It created it in this folder…, see the dll in there…

C:\Windows\ServiceProfiles\NetworkService\AppData\Local\Temp\1508466013\8\0.12.2.1

But errors on the .Convert()

private static IConverter converter = new ThreadSafeConverter(new RemotingToolset<PdfToolset>(new Win64EmbeddedDeployment(new TempFolderDeployment())));

public ActionResult ExportDashboard(List<string> pages)

{

```
     var document = new HtmlToPdfDocument

     {

              GlobalSettings =

             {

                      ProduceOutline = true,

                      DocumentTitle = "Pretty Websites",

                      PaperSize = PaperKind.A4, // Implicit conversion to PechkinPaperSize

                      Margins =

                      {

                               All = 1.375,

                               Unit = Unit.Centimeters

                      }

             }

     };



    //static html

     pages.ForEach(p => document.Objects.Add(new ObjectSettings() { HtmlText = p }));



     byte[] result = converter.Convert(document);



     //handle result



     return Json(true, JsonRequestBehavior.AllowGet);
```

}

**Comment by ghost** [comment]
Hi,

I'm trying this, I hope it helps , greetings from Chile.

``` csharp
private static readonly object _PadLock = new object();
private static IConverter _converter = null;
public static IConverter convertidor {
    get {
        if (_converter == null) {
            lock (_PadLock) {
                if (_converter == null) {
                    _converter = new ThreadSafeConverter(new RemotingToolset<PdfToolset>(new Win32EmbeddedDeployment(new TempFolderDeployment())));
                }
            }
        }
        return _converter;
    }
}
```

then just call the singleton instance

``` csharp
byte[] result = convertidor.Convert(document);

```

**Comment by tuespetre** [comment]
Creating the instance in a static field declaration or in a static constructor would suffice as well.

**Comment by rajeem** [comment]
In my case, I am deploying on a 64-bit VPS then I got this error. I have solved the problem by installing the wkhtmltopdf that I downloaded from http://wkhtmltopdf.org/downloads.html. I chose the 32-bit installer.

**Comment by mtangoo** [comment]
@rajeem that was it

> In my case, I am deploying on a 64-bit VPS then I got this error. I have solved the problem by installing the wkhtmltopdf that I downloaded from http://wkhtmltopdf.org/downloads.html. I chose the 32-bit installer.

**Comment by Dunge** [comment]
I have a similar problem. Using TuesPechkin 2.1.1.0 from NuGet and new StandardConverter(new PdfToolset(new Win64EmbeddedDeployment(new TempFolderDeployment())));

It was working fine on my developer machine, but not when publishing to production:
System.DllNotFoundException: Unable to load DLL 'wkhtmltox.dll': The specified module could not be found. (Exception from HRESULT: 0x8007007E)

The folder containing my binaries have a "TuesPechkin.Wkhtmltox.Win64.dll" file. If I search on disk, I also have a "wkhtmltox.dll" file under C:\Windows\Temp\1698115009\8\0.12.2.1

I did not install the installer package from rajeem comment, I did not remember needing it for my dev machine to work, and would prefer avoid it so that out automatic deployment work on any machine.

Any idea?

**Comment by Dunge** [comment]
Got it, needed VC++ Redist 2013 x64

**Comment by dpcrman** [comment]
Got it, needed VC++ Redist 2013 x32

**Comment by ericsstoll** [comment]
"Got it, needed VC++ Redist 2013 x32."

That fixed for me too.

**Comment by thedumbtechguy** [comment]
"Got it, needed VC++ Redist 2013 x32."

This took me a bit to figure out. Any chance of adding it in the docs?
It only bit me when I deployed to the server.

**Comment by brettandrew** [comment]
I deployed my code to a server today and got the same error everyone was seeing:

`"An error has occurred.","ExceptionMessage":"Unable to load DLL 'wkhtmltox.dll': Access is denied. (Exception from HRESULT: 0x80070005 (E_ACCESSDENIED))","ExceptionType":"System.DllNotFoundException"`

I read a few posts and tried a few things but this is what worked for me in the end and hopefully will for you too.
- Install VC++ Redist 2013 x32 (if you are running 64 bit use x64)
- In my IIS 7 App Pool I had to change the setting LoadUserProfile="True"

Without that last step this TuesPechkin won't be able to write to the temp folder. Hey presto! It worked on the server deployment (finally!). 

Just for completeness here is what I am using for iConverter on an IIS deployment - just an FYI 

``````
  ```
``````

  private static readonly IConverter Converter =
                new ThreadSafeConverter(
                    new RemotingToolset<PdfToolset>(
                        new Win32EmbeddedDeployment(
                            new TempFolderDeployment())));

```
```
